### PR TITLE
Adjustments for SqlMainDomLock and others to make azure operations more resilient

### PR DESCRIPTION
**Pull request by Shazwazza**
Details are here: https://github.com/umbraco/Umbraco-CMS/issues/8215

This should also resolve https://github.com/umbraco/Umbraco-CMS/issues/8392

* Improvements to SqlMainDomLock
  * Better management of db instances and transactions - previously this was trying to manage a single database instances which isn't what should be done, db instances should exist for a short period of time and disposed of. Changed to more explicit code for managing transactions
  * Increase polling times - though according to logs this wasn't causing problems but is still too aggressive for what it's doing
  * Don't stop listening (shut down maindom) if there are SQL exceptions, only if the appdomain has actually triggered a shutdown. We cannot just stop listening even if there are errors since it will lead to premature MainDom shutdown when the app is not actually being shutdown. This means it will just retry on the next poll. 
* Investigation into scheduled publishing lock request timeouts (also reported here #8392) which occurs around the same time as when SqlMainDomLock fails
  * The PerformScheduledPublishInternal was using yield returns with a wrapped Scope/IDisposable which is a bit hard to follow along. One issue is that we were yield returning when the Saving event was canceled for an individual item but not continuing so the logic would have proceeded anyways for that item even if that event was canceled and found other similar issues where it would actually break out of the entire process instead of skip to the next record
  * I believe one of the main issues seen with the lock request timeouts like those in #8392 and in the logs for #8215 is because the PerformScheduledPublishInternal takes a WriteLock at the very beginning of the method. This is the write lock that times out according to the stack traces. This is probably because scheduled publishing runs every minute which is actually a bit crazy to think we lock the whole content tree every minute even if we aren't writing anything. And because of that I think this may also be one of the other main suspects as to why we saw SqlMainDomLock error around the same time we see the sql lock timeout error for scheduled publishing. To fix this:
    * We need a new and much faster method: DocumentRepository.HasContentForRelease instead of only relying on the potentially expensive DocumentRepository.GetContentForRelease, similarly for GetContentForExpiration we should also have HasContentForExpiration. Then we separate the job of PerformScheduledPublishInternal into 2 parts: First, without any read or write locks just check if either HasContentForRelease or GetContentForExpiration  is true, then we can take a write lock and proceed as we are today. This will mean that there isn't a writelock on all content attempted to be made every minute.
* Don't short circuit index rebuilding if one of the populators throws an exception, continue with the other populators
* Brings our sql transient fault handling in-line with the latest specs from MS

### CODE REVIEW

File: src/Umbraco.Core/Services/Implement/ContentService.cs [hunk 1]
```diff
@@ -1387,11 +1453,14 @@ private IEnumerable<PublishResult> PerformScheduledPublishInternal(DateTime date
                             .ToList();
 
                         if (pendingCultures.Count == 0)
-                            break; //shouldn't happen but no point in continuing if there's nothing there
+                            continue; //shouldn't happen but no point in continuing if there's nothing there
```
Comment (clausjensen) [inline src/Umbraco.Core/Services/Implement/ContentService.cs:1456]: maybe update/remove these comments .. someone will be confused in the future if it says no point in continuing and we then `continue` 😃 ..

File: src/Umbraco.Core/Services/Implement/ContentService.cs [hunk 2]
```diff
@@ -1403,94 +1472,51 @@ private IEnumerable<PublishResult> PerformScheduledPublishInternal(DateTime date
...
-                            var impact = CultureImpact.Explicit(culture, IsDefaultCulture(allLangs, culture));
+                            var impact = CultureImpact.Explicit(culture, IsDefaultCulture(allLangs.Value, culture));
                             var tryPublish = d.PublishCulture(impact) && _propertyValidationService.Value.IsPropertyDataValid(d, out invalidProperties, impact);
                             if (invalidProperties != null && invalidProperties.Length > 0)
                                 Logger.Warn<ContentService>("Scheduled publishing will fail for document {DocumentId} and culture {Culture} because of invalid properties {InvalidProperties}",
                                     d.Id, culture, string.Join(",", invalidProperties.Select(x => x.Alias)));
 
                             publishing &= tryPublish; //set the culture to be published
-                            if (!publishing) break; // no point continuing
+                            if (!publishing) continue; // no point continuing
```
Comment (clausjensen) [inline src/Umbraco.Core/Services/Implement/ContentService.cs:1482]: same ^^

File: src/Umbraco.Web/Compose/DatabaseServerRegistrarAndMessengerComponent.cs [hunk 3]
```diff
@@ -218,7 +218,7 @@ public override bool PerformRun()
                 }
                 catch (Exception e)
                 {
-                    _logger.Error<InstructionProcessTask>("Failed (will repeat).", e);
+                    _logger.Error<InstructionProcessTask>(e, "Failed (will repeat).");
```
Comment (clausjensen) [inline src/Umbraco.Web/Compose/DatabaseServerRegistrarAndMessengerComponent.cs:221]: nice catch :)

File: src/Umbraco.Web.UI/Umbraco.Web.UI.csproj [hunk 4]
```diff
@@ -12,7 +12,8 @@
     <ProjectTypeGuids>{349c5851-65df-11da-9384-00065b846f21};{fae04ec0-301f-11d3-bf4b-00c04f79efbc}</ProjectTypeGuids>
     <RunPostBuildEvent>OnBuildSuccess</RunPostBuildEvent>
     <UseIISExpress>true</UseIISExpress>
-    <IISExpressSSLPort>44331</IISExpressSSLPort>
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Web.UI/Umbraco.Web.UI.csproj:15]: Nope not intentional, Visual Studio or something in our git repo constantly changes this everytime I build.
